File: 2018/day1.py
if __name__ == '__main__':
    with open('sampleinputs/day1.txt') as file:
        changes = [int(line if line[0] != '+' else line[1:])
                   for line in file.read().split()]
        part1 = sum(changes)
        print('Part 1: {}'.format(part1))

        frequencies = set()
        frequency = 0
        i = 0
        length = len(changes)
        while True:
            if frequency in frequencies:
                print(i)
                print('Part 2: {}'.format(frequency))
                break
            frequencies.add(frequency)
            frequency += changes[i % length]
            i += 1

File: 2018/Day1.hs
module Day1 where

import Data.Maybe
import qualified Data.Set as Set

parseChange :: (Integral a, Read a) => [Char] -> a
parseChange ('+':xs) = read xs
parseChange xs       = read xs

findDuplicate :: (Num a, Ord a) => [a] -> Maybe a
findDuplicate xs = findDuplicate' Set.empty 0 xs where
    findDuplicate' xset xacc (x:xs) = if xacc `Set.member` xset
                                      then Just xacc
                                      else findDuplicate' (Set.insert xacc xset) (xacc + x) xs

main :: IO ()
main = do
    changes <- map parseChange . lines <$> readFile "inputs/day1.txt"
    let part1 = sum changes
    let part2 = fromJust $ findDuplicate $ cycle changes
    putStrLn $ "Part 1: " ++ show part1
    putStrLn $ "Part 2: " ++ show part2
